storey/flow.py

```diff
@@ -649,6 +649,47 @@ async def _do(self, event):
                 await self._worker_awaitable
 
 
+class Batch(Flow):
+    def __init__(self, max_events, timeout_secs=None, **kwargs):
+        Flow.__init__(self, **kwargs)
+        self._max_events = max_events
+        self._events_count = 0
+        self._batch = []
+        self._timeout_task = None
+
+        self._timeout_secs = timeout_secs
+        if self._timeout_secs <= 0:
+            raise ValueError('Batch timeout cannot be 0 or negative')
+
+    async def sleep_and_emit(self):
+        await asyncio.sleep(self._timeout_secs)
+        await self.emit_batch()
+
+    async def _do(self, event):
+        if event is _termination_obj:
+            self._terminate_worker = True
+            await self.emit_batch()
+            return await self._do_downstream(_termination_obj)
+        else:
+            if len(self._batch) == 0 and self._timeout_secs:
+                self._timeout_task = asyncio.get_running_loop().create_task(self.sleep_and_emit())
+
+            self._events_count = self._events_count + 1
+            self._batch.append(event)
+
+            if self._events_count == self._max_events:
+                await self.emit_batch()
+
+    async def emit_batch(self):
+        if len(self._batch) > 0:
+            await self._do_downstream(Event([self._get_safe_event_or_body(event) for event in self._batch], time=self._batch[0].time))
+            self._batch = []
+            self._events_count = 0
```

Comment:
I think these need to precede the await as well, or you risk another `emit_batch` being triggered before this method resumes.

Reply:
Also, `_events_count` -> `_event_count`